single_kernel_mongo/events/lifecycle.py

```diff
@@ -0,0 +1,154 @@
+    ContainerNotReadyError,
+    UpgradeInProgressError,
+    WorkloadServiceError,
+)
+from single_kernel_mongo.utils.mongo_connection import NotReadyError
+
+logger = logging.getLogger(__name__)
+
+
+class LifecycleEventsHandler(Object):
+    """Events handler for lifecycle events.
+
+    In charge of handling the lifecycle events such as install, start, pebble ready, etc.
+    """
+
+    def __init__(self, dependent: OperatorProtocol, rel_name: str):
+        super().__init__(parent=dependent, key=dependent.name)
+        self.dependent = dependent
+        self.charm = dependent.charm
+        self.relation_name = rel_name
+
+        self.framework.observe(getattr(self.charm.on, "install"), self.on_install)
+        self.framework.observe(getattr(self.charm.on, "start"), self.on_start)
+        self.framework.observe(getattr(self.charm.on, "stop"), self.on_stop)
+        self.framework.observe(getattr(self.charm.on, "leader_elected"), self.on_leader_elected)
+
+        if self.charm.substrate == Substrates.K8S:
+            self.framework.observe(getattr(self.charm.on, "mongod_pebble_ready"), self.on_start)
+
+        self.framework.observe(getattr(self.charm.on, "config_changed"), self.on_config_changed)
+        self.framework.observe(getattr(self.charm.on, "update_status"), self.on_update_status)
+        self.framework.observe(getattr(self.charm.on, "secret_changed"), self.on_secret_changed)
+
+        self.framework.observe(self.charm.on[rel_name].relation_joined, self.on_relation_joined)
+        self.framework.observe(self.charm.on[rel_name].relation_changed, self.on_relation_changed)
+        self.framework.observe(self.charm.on[rel_name].relation_departed, self.on_relation_departed)
+
+        self.framework.observe(
+            getattr(self.charm.on, "mongodb_storage_attached"), self.on_storage_attached
+        )
+        self.framework.observe(
+            getattr(self.charm.on, "mongodb_storage_detaching"), self.on_storage_detaching
+        )
+
+    def on_start(self, event: StartEvent):
+        """Start event."""
+        try:
+            self.dependent.on_start()
+        except Exception as e:
+            logger.error(f"Deferring because of {e}")
+            event.defer()
+            return
+
+    def on_stop(self, event: StopEvent):
+        """Stop event."""
+        self.dependent.on_stop()
+
+    def on_install(self, event: InstallEvent):
+        """Install event."""
+        try:
+            self.dependent.on_install()
+        except (ContainerNotReadyError, WorkloadServiceError):
+            logger.info("Not ready to start.")
+            event.defer()
+            return
+
+    def on_leader_elected(self, event: LeaderElectedEvent):
+        """Leader elected event."""
+        self.dependent.on_leader_elected()
+
+    def on_config_changed(self, event: ConfigChangedEvent):
+        """Config Changed Event."""
+        try:
+            self.dependent.on_config_changed()
```

Comment:
Yes, but as there will be two different operators, this class will call the right methods on the right operator.
This seems to be redundant but it ensures the separation between events and managers so that events never reach the manager class.
The way the `LifeCycleEventsHandler` works is that it will call the `on_config_changed` (for example) method on the provided operator which allows to have one single lifecycle event handler which can be used with different operators.

Reply:
> Where is the logic for this?
This is the inheritance that makes it.
By passing the correct `dependent` `Operator` when creating the even handler, it's its methods that will be called.